### PR TITLE
Allow hauling items along the ground

### PR DESCRIPTION
**Pull request by anonym**
#### Summary
SUMMARY: Features "Allow hauling items along the ground."

#### Purpose of change
Without the proposed hauling feature the closest thing we have is Advanced Inventory and its "Move all items", but it is painful to use for large distances and complex paths with many twists and turns.

#### Describe the solution
The new haul action is toggleable; while enabled all items on the player's position are moved ("hauled") along with the player using the `ACT_MOVE_ITEMS` activity so we get proper move cost, it can be interrupted, etc.

### CODE REVIEW

**Comment by kevingranade** [comment]
So you're modeling having the player repeatedly pick up and drop all the items while moving?
What does it do when you are hauling things, and enter a square that already has items?
What happens if the player is attacked?
Enters a vehicle?

**Comment by anonym** [comment]
> So you're modeling having the player repeatedly pick up and drop all the items while moving?
Yes, it literally automates repeated use of the "Move all items" from Advanced Inventory management, which in my experience is how players move big piles, e.g. newly chopped tree logs.

> What does it do when you are hauling things, and enter a square that already has items?
The new items are now also subject to hauling. I like how this can be used to quickly collect piles into one.

> What happens if the player is attacked?
Since it's implemented as an activity (`ACT_MOVE_ITEMS`, same as Advanced Inventory management uses) the hauling will be interrupted with the usual prompt when attacked (or safe mode triggers etc).

> Enters a vehicle?
Vehicle source and target tiles are not supported, so hauling cannot be initiated on such tiles, and will be automatically aborted before entering such tiles. Same applies to deep water tiles, tiles on different z-levels, and tiles you can move onto but that cannot store items (via `map.can_put_items()`, so e.g. rubble, opened windows).

**Comment by anonym** [comment]
The Travis CI build seems to have [stalled](https://travis-ci.org/CleverRaven/Cataclysm-DDA/jobs/432156766) and not genuinely failed.

**Comment by anonym** [comment]
> > So you're modeling having the player repeatedly pick up and drop all the items while moving?
>
> Yes, it literally automates repeated use of the "Move all items" from Advanced Inventory management [...]

It turns out I made hauling inferior to repeated "Move all items"! Hauling always does one square moves, but with Advanced Inventory management you can do two-square moves (e.g. East to West) for the same move cost, so it's twice as time efficient at moving items! I've pushed a commit that factors distance into the move cost, eliminating this inconsistency.

File: src/handle_action.cpp [hunk 1]
```diff
@@ -534,6 +534,29 @@ static void grab()
...
+    if( u.is_hauling() ) {
+        u.stop_hauling();
+    } else {
+        if( m.veh_at( u.pos() ) ) {
+            add_msg( m_info, _( "You cannot haul inside vehicles." ) );
+        } else if( m.has_flag( TFLAG_DEEP_WATER, u.pos() ) ) {
+            add_msg( m_info, _( "You cannot haul while in deep water." ) );
+        } else if( !m.can_put_items( u.pos() ) ) {
+            add_msg( m_info, _( "You cannot haul items here." ) );
+        } else if( ! m.has_items( u.pos() ) ) {
```
Comment (BevapDin) [inline src/handle_action.cpp]: Inconsistent style: there is a space in front of the `!`.

File: src/player.cpp [hunk 2]
```diff
@@ -10763,6 +10767,9 @@ bool player::has_activity(const activity_id type) const
 
 void player::cancel_activity()
 {
+    if ( has_activity( activity_id( "ACT_MOVE_ITEMS" ) ) && is_hauling() ) {
```
Comment (BevapDin) [inline src/player.cpp]: Inconsistent style: here is a space after the `if`.

File: src/player.cpp [hunk 3]
```diff
@@ -11169,6 +11176,29 @@ object_type player::get_grab_type() const
...
+    if ( is_armed() ) {
+        add_msg( m_warning, _( "Your hands are not free, which makes hauling slower." ) );
+    }
+    this->hauling = true;
+}
+
+void player::stop_hauling()
+{
+    add_msg( _( "You stop hauling items." ) );
+    this->hauling = false;
```
Comment (BevapDin) [inline src/player.cpp]: Why are you accessing `hauling` via `this` instead of directly?

Comment (anonym) [reply src/player.cpp]: Good  question! Python brain damage? :)

**Comment by anonym** [comment]
@BevapDin, thanks again for linting me! I didn't have @astyle@ installed initially but now do so I should be causing fewer silly mistakes like this.

File: src/advanced_inv.cpp [hunk 4]
```diff
@@ -2407,28 +2432,14 @@ void advanced_inv_area::set_container_position()
             off = g->u.grab_point;
             break;
         case AIM_SOUTHWEST:
```
Comment (ZhilkinSerg) [inline src/advanced_inv.cpp]: I guess you can remove all cases except `AIM_DRAGGED` as they are still checked in new `aim_vector` function.

File: src/advanced_inv.cpp [hunk 5]
```diff
@@ -2407,28 +2432,14 @@ void advanced_inv_area::set_container_position()
...
-            break;
         case AIM_NORTHWEST:
-            off = tripoint( -1, -1, 0 );
-            break;
         case AIM_NORTH:
-            off = tripoint( 0, -1, 0 );
-            break;
         case AIM_NORTHEAST:
-            off = tripoint( 1, -1, 0 );
+            off = aim_vector( ( aim_location )uistate.adv_inv_container_location );
```
Comment (ZhilkinSerg) [inline src/advanced_inv.cpp]: Please don't use C-style casts. Use `static_cast<aim_location>` instead.

File: src/game.cpp [hunk 6]
```diff
@@ -10604,6 +10605,24 @@ bool game::walk_move( const tripoint &dest_loc )
         }
     }
 
+    if( u.is_hauling() ) {
+        u.assign_activity( activity_id( "ACT_MOVE_ITEMS" ) );
+        u.activity.values.push_back( false );
+        u.activity.values.push_back( false );
```
Comment (ZhilkinSerg) [inline src/game.cpp]: Duplicate line. Is it intended?

File: src/advanced_inv.cpp (hunk 4 above)
Comment (anonym) [reply src/advanced_inv.cpp]: Good point! Fixed!

File: src/game.cpp (hunk 6 above)
Comment (anonym) [reply src/game.cpp]: Yes, I added comments explaining what's going on.

File: src/advanced_inv.cpp (hunk 5 above)
Comment (anonym) [reply src/advanced_inv.cpp]: Ack.

**Comment by alanbrady** [comment]
Hrm this can lead to move starvation https://github.com/CleverRaven/Cataclysm-DDA/pull/25796/files#diff-570f284881dfa11d0f06697a7e5fe6dcR1704

**Comment by anonym** [comment]
> Hrm this can lead to move starvation https://github.com/CleverRaven/Cataclysm-DDA/pull/25796/files#diff-570f284881dfa11d0f06697a7e5fe6dcR1704

Interesting, could you please elaborate?

**Comment by alanbrady** [comment]
So if you just subtract a bunch of moves like you've done here, the game will just eat up turns until you have a positive number and can actually do something.

This means if you subtract a ton of moves it leaves the player vulnerable to zombie mobbing and other problems.

In the cases where we need to potentially use multiple turns worth of moves, we have a whole activity_handler infrastucture to do that in a way that the player can interrupt it and take care of the hulk that just came into view instead of getting pulped into oblivion because the player wanted to haul something.

I noticed a second line where you're doing this in this patch as well.

**Comment by anonym** [comment]
I get your concern! Note, though, that `distance` will either be `1` (same as before) or `2`, so it will at most be twice as bad as before, so I don't think this will have a noticeable effect in practice (during my tests I never noticed anything).

However, my other merged PR (5daaf031f59032c06da99d0f2d4156d5b40fada7) combined with your concern seems like a bad combo. So instead of calculating:

    move_cost = ( by_charges ? 1 : amount_to_move ) * Pickup::cost_to_move_item( g->u, new_item );

and later `g->u.mod_moves()` for the total amounts of moves, we should call it repeatedly for each item, i.e.:

    for( int i = 0; i < ( by_charges ? 1 : amount_to_move ); i++) {
      // <insert code for moving one item>
      g->u.mod_moves( -distance * Pickup::cost_to_move_item( g->u, new_item ) );
    }

Right? If so I'll open a new PR.

**Comment by alanbrady** [comment]
>  for the total amounts of moves, we should call it repeatedly for each item,

Not quite.  Take a look at acitivity_handlers.cpp and the places where we use them to get a better idea of how to do it.

**Comment by AdonaiJr** [comment]
Tnx so much for this. Closes my original request #23594

**Pull request merged**